addons/web_editor/static/src/js/backend/convert_inline.js

```diff
@@ -697,42 +705,20 @@ async function toInline($editable, cssRules, $iframe) {
     $editable.addClass('odoo-editor-editable');
 }
 /**
- * Take all elements with a `background-image` style and convert them, along
- * with their contents, to `png` images. The images are then appended to the
- * original elements stripped of their background images and padding, while all
- * of their children are removed.
+ * Take all elements with a `background-image` style and convert them to `vml`
+ * for Outlook.
  *
  * @param {Element} editable
  */
-async function flattenBackgroundImages(editable) {
-    for (const backgroundImage of editable.querySelectorAll('*[style*=background-image]')) {
-        if (backgroundImage.parentElement) { // If the image was nested, we removed it already.
-            const iframe = document.createElement('iframe');
-            const style = getComputedStyle(backgroundImage);
-            const clonedBackground = backgroundImage.cloneNode(true);
-            clonedBackground.style.height = style['height'];
-            clonedBackground.style.width = style['width'];
-            iframe.style.height = style['height'];
-            iframe.style.width = style['width'];
-            backgroundImage.after(iframe);
-            iframe.contentDocument.body.append(clonedBackground);
-            iframe.contentDocument.body.style.margin = 0;
-
-            const canvas = await html2canvas(clonedBackground, { scale: 1 });
-            const image = document.createElement('img');
-            image.setAttribute('src', canvas.toDataURL('png'));
-            image.setAttribute('width', canvas.getAttribute('width'));
-            image.setAttribute('height', canvas.getAttribute('height'));
-            image.style.setProperty('margin', 0);
-            image.style.setProperty('display', 'block'); // Ensure no added vertical space.
-            // Clean up the original element.
-            backgroundImage.replaceChildren();
-            backgroundImage.style.setProperty('padding', 0);
-            backgroundImage.style.removeProperty('background-image');
-            backgroundImage.prepend(image); // Add the image to the original element.
-            backgroundImage.className = backgroundImage.className.replace(/p[bt]\d+/g, ''); // Remove padding classes.
-            iframe.remove();
-        }
+function flattenBackgroundImages(editable) {
+    const backgroundImages = [...editable.querySelectorAll('*[style*=background-image]')]
+        .filter(el => el.style.backgroundImage.includes('url(') && !el.closest('.mso-hide'))
+        .reverse();
+    for (const backgroundImage of backgroundImages) {
+        // Put the Outlook version after the original one in an mso conditional.
+        backgroundImage.after(_createMso(_backgroundImageToVml(backgroundImage)));
+        // Hide the original element for Outlook.
+        backgroundImage.classList.toggle('mso-hide', true);
```

Comment:
Now that I think about it, why use a forced toggle instead of `backgroundImage.classList.add('mso-hide')` though ? The only case where a forced toggle makes sense is when you have a reference to a boolean variable and you don't want to do `if (myvar) then add(class) else remove(class)`, but otherwise I don't really see the point.

Reply:
Wait what ? Classlist doesn't work like that. If the class is already present then it won't do anything.